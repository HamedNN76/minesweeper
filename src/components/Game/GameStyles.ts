import {makeStyles} from '@mui/styles';

export const gameStyles = makeStyles({
  cell: {
    borderRadius: '0px !important',
    padding: '8px !important',
    height: 32,
    width: 32,
    minWidth: '32px !important',
  },
  text: {
    fontWeight: 'bold',
    margin: 0,
  },
  row: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
});
