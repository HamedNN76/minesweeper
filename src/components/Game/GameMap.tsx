import React, {useMemo} from 'react';
import {Button, ButtonProps} from '@mui/material';
import {gameStyles} from './GameStyles';

export type GameMapProps = {
  square: string;
  onClick: ButtonProps['onClick'];
  testId: string;
};

export function GameMap(props: GameMapProps) {
  const {square, onClick, testId} = props;
  const classes = gameStyles();

  const color = useMemo<ButtonProps['color']>(() => {
    if (square === '□') {
      return 'primary';
    }
    if (square === '*') {
      return 'error';
    }
    return 'success';
  }, [square]);

  return (
    <Button
      variant="outlined"
      color={color}
      onClick={onClick}
      className={classes.cell}
      data-testid={testId}
    >
      <span className={classes.text}>{square !== '□' ? square : ''}</span>
    </Button>
  );
}
