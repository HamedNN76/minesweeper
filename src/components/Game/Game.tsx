import React, {useCallback} from 'react';
import {GameSocket} from '../../utils/GameSocket';
import {gameStyles} from './GameStyles';
import {GameMap} from './GameMap';

interface Props {
  gameMap: string[];
}

export function Game({gameMap}: Props) {
  const classes = gameStyles();

  const onCellClick = useCallback((y: number, x: number) => {
    GameSocket.socket.send(`open ${x} ${y}`);
  }, []);

  const renderMap = useCallback(
    (gameMap: string[]) => {
      return gameMap.map((game: string, row: number) => {
        const squares = game.split('');
        const rowItems = squares.map((square: string, column: number) => {
          const key = `square-${row}-${column}`;
          const testId = `square-${row}-${column}`;
          return (
            <GameMap
              key={key}
              square={square}
              onClick={() => onCellClick(row, column)}
              testId={testId}
            />
          );
        });

        return (
          <div className={classes.row} key={`square-row-${row}`}>
            {rowItems}
          </div>
        );
      });
    },
    [classes.row, onCellClick],
  );

  if (!gameMap.length) {
    return <p>Choose a level and click on Play</p>;
  }

  return <>{renderMap(gameMap)}</>;
}
