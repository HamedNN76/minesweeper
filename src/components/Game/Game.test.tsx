import React from 'react';
import renderer from 'react-test-renderer';
import {Game} from './Game';
import {mount} from 'enzyme';
import {cleanup} from '@testing-library/react';

describe('<Game />', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });
  it('Match Game to SnapShot', () => {
    const gameMap: string[] = [];
    const tree = renderer
        .create(
            <Game gameMap={gameMap}/>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Should find cell 1-1', () => {
    const gameMap: string[] = [
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
    ];
    const wrapper = mount(
        <Game gameMap={gameMap}/>,
    );
    const element = wrapper.find('square-1-1');
    expect(element).toBeTruthy();
  });
});
