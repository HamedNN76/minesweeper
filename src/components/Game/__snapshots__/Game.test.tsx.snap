// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> Match Game to SnapShot 1`] = `
<p>
  Choose a level and click on Play
</p>
`;
